Add documentation for testing and CI (#3532)
ethereum/web3.js: TESTING.md

| windows | -- | [e2e.windows.sh][5] | Verifies Web3 installs on Windows OS / Node 12 and can connect to Infura over wss and https | :white_check_mark: |

## Optional Tests

CI also has tests that install Web3's state at an arbitrary commit in an external real-world project and run *their* unit tests with it. This strategy is borrowed from ethereum/solidity which checks latest Solidity against OpenZeppelin and others to keep abreast of how local changes might affect critical projects downstream from them.

Examples include:
+ [e2e.mosaic.sh][8]: ~300 unit tests for [a Solidity project built with Buidler & @truffle/contract][9]
+ [e2e.ganache.core.sh][9]: ~600 unit tests for [a widely used JS testrpc][11]

These tests are "allowed failures". They're:
+ a pre-publication sanity check that discovers how Web3 performs in the wild
+ useful for catching problems which are difficult to anticipate
+ exposed to failure for reasons outside of Web3's control, ex: when fixes here surface bugs in the target.

## Implementation Details

**Code coverage**

Coverage is measured by aggregating the results of tests run in the `unit_and_e2e_clients`
CI job.

**Tests which use an Ethereum client**

> The npm script `test:e2e:clients` greps all tests with an `[ E2E ]` tag

Author: ```suggestion
The npm script `test:e2e:clients` greps all tests with an `[ @E2E ]` tag
```

in their mocha test description and runs them against:
+ ganache-cli

> + geth 1.9.13 (POA, single instance, instamining)

Reviewer: let's make sure to update this file if we end up moving the pin, i could see it getting confusing if out of date

Author: Yeah, good catch. I'll add a note in that issue.

+ geth 1.9.13 (POA, single instance, mining at 2s intervals)

These tests are grouped in files prefixed by "e2e", ex: `test/e2e.method.call.js`.

Additionally, there are conventional unit tests postfixed `.ganache.js` which spin up a ganache
server programatically within mocha. This pattern is useful if you want to
control client configuration for a specific case, test against multiple independent providers, etc.

**"Real world" tests**

The tests which install Web3's current state in an external real-world project and
run their unit tests accomplish this by publishing the monorepo to an ephemeral private
npm registry which is spun up in CI using [verdaccio][14]. (Implementation details can
be seen in [scripts/e2e.npm.publish.sh][15])

> The real world target is then cloned and npm or yarn are used to replace the target's existing

Author: ```suggestion
The real world target is then cloned and npm or yarn are used to replace its existing
```

Web3 version with the version published to the the private registry. A simple example can be seen at
[scripts/e2e.ganache.core.sh][10].

In practice, complex projects can have many versions of Web3 nested in their dependency tree.

> For the test to be valid, it's important to coerce all of them to the virtually published package.

Author: ```suggestion
It's important to coerce all of them to the virtually published package's version for the test to be valid.
```
